ci(aio): aio-builds-setup improvements

gkalpak opened the pull request:
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


**What kind of change does this PR introduce?** (check one with "x")
```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[x] CI related changes
[ ] Other... Please describe:
```

**Does this PR introduce a breaking change?** (check one with "x")
```
[ ] Yes
[x] No
```


**Other information**:
Some small improvements and docs for `aio-builds-setup`.
Partly addresses #14922.

gkalpak committed:
build(aio): remove redundant arg

gkalpak committed:
feat(aio): redirect HTTP to HTTPS

gkalpak committed:
feat(aio): make it easy to keep relevant logs outside the docker container

mary-poppins commented:
The angular.io preview for 4185be5 is available [here][1].

[1]: https://pr15012-4185be5a68e2cf4ac434629563b4276fb94ae12d.ngbuilds.io/

mary-poppins commented:
The angular.io preview for 0d33532 is available [here][1].

[1]: https://pr15012-0d33532e669444528250a8b407436fdf7f0e461d.ngbuilds.io/

mary-poppins commented:
The angular.io preview for f6cadec is available [here][1].

[1]: https://pr15012-f6cadec64f660bfc96df4fdeb8e774c53ee4b12f.ngbuilds.io/

mary-poppins commented:
The angular.io preview for b25bdfa is available [here][1].

[1]: https://pr15012-b25bdfa7f31f645ac8faa7db26520e3c36d3d493.ngbuilds.io/

gkalpak commented:
This is now ready for review :smiley:

gkalpak committed:
docs(aio): add more docs about `aio-builds-setup`

gkalpak committed:
ci(aio): fix clean-up script on `ngbuilds.io`

mary-poppins commented:
The angular.io preview for 39cd3ce is available [here][1].

[1]: https://pr15012-39cd3cedf68aee70d53dadf7b58b4fbc8342671a.ngbuilds.io/

mary-poppins commented:
The angular.io preview for b916c26 is available [here][1].

[1]: https://pr15012-b916c26e89115662e84567d68f6a686cdfbf33bb.ngbuilds.io/

gkalpak commented:
Actually, the 5th commit (fixing the clean-up script is not sufficient). Working on it.
The first 4 commits are good to review.

mary-poppins commented:
The angular.io preview for 2b553de is available [here][1].

[1]: https://pr15012-2b553de731761997fa95591152e8c2b0a1297e6a.ngbuilds.io/

mary-poppins commented:
The angular.io preview for cb0fada is available [here][1].

[1]: https://pr15012-cb0fada85cf4f98d70559951e1211f48a69e7952.ngbuilds.io/

mary-poppins commented:
The angular.io preview for 385e2e9 is available [here][1].

[1]: https://pr15012-385e2e951f78822d06f0f1356446804ddbc418b1.ngbuilds.io/

gkalpak commented:
Fixed the clean-up script.
This is now (very, very) ready for review :smiley:

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/nginx/aio-builds.conf:
```diff
@@ -1,9 +1,19 @@
+# Redirect all HTTP traffic to HTTPS
+server {
+  server_name _;
```
Should we be using this approach?
See https://blog.gahooa.com/2013/08/21/nginx-how-to-specify-a-default-server/

gkalpak replied:
We are already using `default_server` as suggested in the blog post. `server_name _` is not about making it a default server; it is just a wildcard, so we didn't have to specify the server name (e.g. domain name).

This was useful for making the docker setup more easily reusable by other projects (because at first there was no need for specifying the domain name). It is not likely that this particular container will ever serve for multiple unrelated domain names, but since we now already have the doain name as an env var (for other purposes) we can use it in this config too.

gkalpak replied:
Actually, due to the precedence wildcard server names get over regexps (see [here][1], it will unnecessarily complicate our config. So, I will leave it as is.

[1]: http://nginx.org/en/docs/http/server_names.html

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/nginx/aio-builds.conf:
```diff
@@ -1,9 +1,19 @@
+  listen [::]:{{$AIO_NGINX_PORT_HTTP}};
+
+  # Ideally we want 308 (permanent + keep original method),
+  # but it is relatively new and not supported by some clients (e.g. cURL).
```
why do we care about cURL in this scenario?
If we are making cURL requests then it ought to be to the correct protocol.

petebacondarwin replied:
Oh I see, you are using cURL for testing... is that the use case?

gkalpak replied:
Yes, I am using cURL for tests. Also, I don't know what other browsers might not support it properly (people are using all sorts of unknown-to-me browsers all over the world).

I am fine going either way.

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-js/lib/verify-setup/nginx.e2e.ts:
```diff
@@ -3,230 +3,250 @@ import * as path from 'path';
-    beforeEach(() => {
-      h.createDummyBuild(pr, sha9);
-      h.createDummyBuild(pr, sha0);
+    h.runCmd(`curl -iL ${httpUrl}`).then(result => {
```
why not just check that we get the correct HTTP response code?

gkalpak replied:
Yeah, that would work as well :smiley:
Although, it would be a good idea to also test the redirected URL too.

gkalpak replied:
Done.

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-js/lib/verify-setup/nginx.e2e.ts:
```diff
@@ -3,230 +3,250 @@ import * as path from 'path';
+      it('should respond with 403 for directories', done => {
+        Promise.all([
+          h.runCmd(`curl -iL ${scheme}://pr${pr}-${sha9}.${host}/foo/`).then(h.verifyResponse(403)),
+          h.runCmd(`curl -iL ${scheme}://pr${pr}-${sha9}.${host}/foo`).then(h.verifyResponse(403)),
```
I once read somewhere that returning 403 was less "safe" than returning 404, since it gives an attacker more information than they would have otherwise. E.g. in this case, that the URL might be a directory?

I guess it probably doesn't matter...

gkalpak replied:
That's what nginx does by default. I don't think it matters for thsi setup.

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-js/lib/verify-setup/nginx.e2e.ts:
```diff
@@ -3,230 +3,250 @@ import * as path from 'path';
-        h.runCmd(`curl -iL ${scheme}://${host}/health-check/`).then(h.verifyResponse(200)),
-      ]).then(done);
-    });
+      it('should accept SHAs with leading zeros (but not ignore them)', done => {
```
what does it mean to not ignore a request in this case?
the body of the test indicates that the leading 0 should cause the SHA request to be rejected?

gkalpak replied:
This test description is a little confusing. I mean "not ignore the zeros". So, `SHA: 0123` should not be treated as `123`.
But leading zeros are valid in SHAs (in contrast to what happens for PRs, where leading zeros are not valid, because PR numbers don't have leading zeros).

gkalpak replied:
I updated the description to `...(but not trim the zeros)` and added an extra expectation for `sha9` (without the leading `0`). Hopefully that makes it a little more obvious.

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-js/lib/verify-setup/nginx.e2e.ts:
```diff
@@ -3,230 +3,250 @@ import * as path from 'path';
-      ]).then(done);
-    });
+      it('should accept SHAs with leading zeros (but not ignore them)', done => {
+        const bodyegex = new RegExp(`^PR: ${pr} | SHA: ${sha0} | File: /index\\.html$`);
```
body_R_egex?

gkalpak replied:
Oh, I have fixed it in #15049: https://github.com/angular/angular/pull/15049/files#diff-a1b999efd1172f9cf7ed6c81bb1313ffL96

gkalpak committed:
test(aio): improve test description and expectations

petebacondarwin commented on aio/aio-builds-setup/docs/image-config--environment-variables.md:
```diff
@@ -0,0 +1,52 @@
+you don't need to specify values for those.
+
+- `AIO_BUILDS_DIR`:
+  The directory (inside the container) where the uploaded build artifacts are kept.
```
artifact :-) (Americanism)

gkalpak replied:
I looked it up before using it (because although I have learned it is `artifact` in my english lessons, I've seen many people using `artefact`). :smiley:
According to [this][1]:

> The British preference for artefact is a new development.

So, I went old-school 😛 

[1]: http://grammarist.com/spelling/artefact-artifact/

petebacondarwin commented on aio/aio-builds-setup/docs/overview--general.md:
```diff
@@ -0,0 +1,84 @@
+
+## Security model
+Since we are managing a public server, it is important to take appropriate measures in order to
+prevent abuse. For more details on the chanllenges and the chosen approach take a look at the
```
typo in challenges

gkalpak replied:
Fixed.

petebacondarwin commented on aio/aio-builds-setup/docs/overview--general.md:
```diff
@@ -0,0 +1,84 @@
+### Removing obsolete artifacts
+In order to avoid flooding the disk with unnecessary build artifacts, there is a cronjob that runs a
+clean-up tasks once a day. The task retrieves all open PRs from GitHub and removes all directories
+that do not correspond with an ppen PR.
```
typo in open

gkalpak replied:
Fixed.

petebacondarwin reviewed (approved):
Loving the docs!

petebacondarwin commented:
Feel free to do any last git cleanup and then apply the merge label.

gkalpak committed:
build(aio): replace all occurrences of env vars on a line in `aio-builds-setup`

mary-poppins commented:
The angular.io preview for 68c534b is available [here][1].

[1]: https://pr15012-68c534b187bb86bcd661972ce9dfddd97c26dcb4.ngbuilds.io/

mary-poppins commented:
The angular.io preview for 0588ee9 is available [here][1].

[1]: https://pr15012-0588ee97429f56cdf78ff9e3c9a76fee9e6851a8.ngbuilds.io/

mary-poppins commented:
The angular.io preview for 7310b14 is available [here][1].

[1]: https://pr15012-7310b14a2cbbf7ebd2d27735f2af8fd9e59675a7.ngbuilds.io/

gkalpak commented:
I rebased on master (just in case) and addressed the comments.
Added 5 new commits (one addresses a comment and another one fixes a potential future bug :smiley:)

@petebacondarwin can you take a quick look please?

mary-poppins commented:
The angular.io preview for f24daf7 is available [here][1].

[1]: https://pr15012-f24daf718f09e47de8a6042d17e0c4518e7a9864.ngbuilds.io/

mary-poppins commented:
The angular.io preview for 5387a47 is available [here][1].

[1]: https://pr15012-5387a47a3422dbaff141ad7204aee25c29e8364d.ngbuilds.io/

petebacondarwin commented:
5 new commits??
Perhaps only two?

petebacondarwin reviewed (approved):
still LGTM

gkalpak commented:
Right. They were 5, I swear (but then I decided to squash 3 of them into previous ones and forgot to update the comment) :smiley:

chuckjaz commented:
This needs to be rebase

gkalpak commented:
@chuckjaz, rebased.

mary-poppins commented:
The angular.io preview for 5d495ed is available [here][1].

[1]: https://pr15012-5d495edab3e862f39074170bae8b3da4276bc1b8.ngbuilds.io/
